### PR TITLE
Reverb fix #144

### LINKED ISSUE #144: Reverb causes Cracking
With the following Mapping I got some very bad sound quality (especially from the middle of the playback onward). I'm not sure what causes this though

- Dates from 2000 to 2023
- Audio Length = 2:30
- Vodafone: Price -> Retro Synth: Pitch
- Exxon: Price      -> Retro Synth: Relvolume
- Meta: Price       -> Retro Synth: Delay Echo
- Alibaba: Price   -> Retro Synth: Feedback Echo
- Apple: Price     -> Retro Synth: Delay Reverb

(Note: This were tested after my fix of #143, bc it crashed otherwise)
(Note: I also tested this a couple of times, and the sound always had the same bad quality)

### PR DESCRIPTION
**Pull request by Rex2002**
please take a look that, if you find something that looks somehow weird, please report. 
Somewhere in there is a bug still hiding I believe which causes occasional cracks (but nothing showstopper-like, only occurs around once for ~0.25s in every third 2.30min sonification).

### CODE REVIEW

File: src/main/java/audio/Util.java [hunk 1]
```diff
@@ -58,4 +58,8 @@ public static byte[] convertShortToByte(short[] data) {
         }
         return outBuffer;
     }
+
+    public static double findAverage(double[] preOut) {
```
Comment (MalteRichert) [inline src/main/java/audio/Util.java:62]: This does the same as Maths.mean() that was just added in #160 . One of them should be removed, right?

File: src/main/java/audio/harmonizer/Harmonizer.java [hunk 2]
```diff
@@ -151,18 +151,40 @@ private int[] normalizeDelayEcho() throws AppError {
...
+                int[] output = new int[delayEcho.length];
+                for (int i = 0; i < output.length; i++) {
+                    double delayEchoVal = delayEcho[i];
+                    checkDouble(delayEchoVal, "delayEcho", i);
+                    output[i] = delayEchoVal == -1 ? defaultDelay : (int) (delays[(int) (delayEchoVal * (delays.length - 1))] * SAMPLES_PER_BAR);
+                }
+                return output;
+            }
+            else{
+                System.out.println("max no delay values: " + maxNoDelayValues);
```
Comment (MalteRichert) [inline src/main/java/audio/harmonizer/Harmonizer.java:169]: We probably want to remove prints at some point

File: src/main/java/audio/harmonizer/Harmonizer.java [hunk 3]
```diff
@@ -207,14 +229,36 @@ private int[] normalizeDelayReverb() throws AppError {
...
         int DEFAULT_DELAY_REVERB = Constants.SAMPLE_RATE / 25; // number of samples corresponding to 40ms
         double[] delayReverb = dataRaw.getDelayReverb();
-        if (delayReverb != null) {
-            int[] output = new int[delayReverb.length];
-            for (int i = 0; i < delayReverb.length; i++) {
-                checkDouble(delayReverb[i], "delayReverb", i);
+        int maxNoDelayValues = (int) (1.0/3 * (numberBeats / (Constants.TEMPO / 60f)));
 
-                output[i] = delayReverb[i] == -1 ? DEFAULT_DELAY_REVERB : (int) (delayReverb[i] * MAX_DELAY_REVERB);
+        if (delayReverb != null) {
```
Comment (MalteRichert) [inline src/main/java/audio/harmonizer/Harmonizer.java:234]: This looks like a lot of code that's very similar to the echo code, can we factor it out somehow?

File: src/main/java/audio/synth/fx/Effect.java [hunk 4]
```diff
@@ -49,32 +50,72 @@ public static double[] echoWithOverdrive(double[] input, double feedback, int de
...
+     * This is achieved by scaling down the feedback to zero, each time the delay value changes using the envelope.
+     * Since the scaling of the feedback can not be done to fast, because then the cracking reappears,
+     * this method is limited to delay arrays that have at most a length of 1/3 of the sound length in seconds.
+     * (the length limiting of the delayArray is already done in the Harmonizer)
+     * @param input the sound-array that is supposed to be echoed / reverbed
+     * @param feedback the feedback values that are to be applied
+     * @param delayArray the delay values that are to be applied
+     * @return the input with respective reverb / echo effect
+     */
+    public static double[] echoWithFeeeeedback(double[] input, double[] feedback, int[] delayArray){
```
Comment (MalteRichert) [inline src/main/java/audio/synth/fx/Effect.java:67]: I think there's a drop of energy drink stuck in the naming of this method xd

**Review by MalteRichert — APPROVED** [review]
Love the effort you put into this! The remaining crackling is probably my fault from the BOOM event inst

**Review by ArtInLines — APPROVED** [review]
🎉 Damn, you did it. I'm genuinely astonished how much better it works now 👍 Amazing job

**Comment by MalteRichert** [comment]
@Rex2002 I really don't want to break anything but I think this should do the delay normalisation just the same way with half as much code. There's only two places where reverb and echo behave differently and they are solved with a simple if-else.

**Review by Rex2002 — COMMENTED** [review]
yes, that indeed looks cleaner and does not add as much complexity as I expected, very cool :+1:

**Pull request merged**